core/src/main/scala/org/apache/spark/sql/execution/columnar/impl/ColumnFormatEntry.scala

```diff
@@ -294,80 +331,249 @@ class ColumnFormatValue extends SerializedDiskBuffer
               columnBuffer = v.columnBuffer
+              decompressionState = v.decompressionState
+              fromDisk = true
               // restart reference count from 1
-              while (true) {
-                val refCount = SerializedDiskBuffer.refCountUpdate.get(this)
-                val updatedRefCount = if (refCount <= 0) 1 else refCount + 1
-                if (SerializedDiskBuffer.refCountUpdate.compareAndSet(
-                  this, refCount, updatedRefCount)) {
-                  return duplicateBuffer(columnBuffer)
-                }
-              }
-            }
+              refCount = 1
+              return transformValueRetain(columnBuffer, decompress, compress)
+
             case null | _: Token => // return empty buffer
             case o => throw new IllegalStateException(
               s"unexpected value in column store $o")
           }
-        } catch {
-          case _: EntryDestroyedException | _: DiskAccessException |
-               _: RegionDestroyedException =>
-            // These exception types mean that value has disappeared from disk
-            // due to compaction or bucket has moved so return empty value.
-            // RegionDestroyedException is also ignored since background
-            // processors like gateway event processors will not expect it.
         }
+      } catch {
+        case _: EntryDestroyedException | _: DiskAccessException |
+             _: RegionDestroyedException =>
+        // These exception types mean that value has disappeared from disk
+        // due to compaction or bucket has moved so return empty value.
+        // RegionDestroyedException is also ignored since background
+        // processors like gateway event processors will not expect it.
+      } finally {
+        dr.releaseReadLock()
       }
-      DiskEntry.Helper.NULL_BUFFER.duplicate()
     }
+    this
   }
 
-  override final def needsRelease: Boolean = columnBuffer.isDirect
+  private def transformValueRetain(buffer: ByteBuffer, decompress: Boolean,
+      compress: Boolean): ColumnFormatValue = {
+    val result = transformValue(buffer, decompress, compress)
+    if (result ne this) {
+      // decrement reference count that has been incremented by caller
+      assert(decrementReference())
+    }
+    result
+  }
 
-  override protected def releaseBuffer(): Unit = synchronized {
+  @inline private def getCachePerfStats(context: RegionEntryContext): CachePerfStats = {
+    if (context ne null) context.getCachePerfStats
+    else GemFireCacheImpl.getExisting.getCachePerfStats
+  }
+
+  private def transformValue(buffer: ByteBuffer, decompress: Boolean,
+      compress: Boolean): ColumnFormatValue = {
+    if (decompress) {
+      if (this.decompressionState != 0) {
+        if (this.decompressionState > 1) {
+          this.decompressionState = 1
+        }
+        this
+      } else {
+        // decompress and replace buffer if possible
+        val isDirect = buffer.isDirect
```

Comment:
Why direct buffers are not compressed ? Underlying compressors does not support ?

Reply:
Actually direct buffers work best. This check is that if its a direct buffer and someone else is using the underlying buffer data (this.refCount > 2), then skip changing the buffer since the direct buffer should be released right away. For a heap buffer, the user can continue using it and will be handled by GC (but has the problem of inflated and unaccounted memory usage till the time other references are alive which is a general problem with heap buffers as of now until HeapBufferAllocator also does not take care of tracking the usage like ManagedDirectBufferAllocator).